### ISSUE TITLE
Strict type tokens for other theme variables

### ISSUE THREAD

**Issue by lexanth** [chakra-ui/chakra-ui#5520]
### Description

https://github.com/chakra-ui/chakra-ui/pull/5243 added support for generating strict types for variants and sizes

I'd like to be able to extend that to other properties (e.g. `colors`, `space` etc.)

### Problem Statement/Justification

We'd like to ensure that only known colours/sizes etc are used in the app to ensure consistency.

For example, `colors` supports `| (string & {})`, so any colour can be provided. This includes HTML named colours, which make it hard to tell from looking at some code whether it is using known theme values or just arbitrary values.
```tsx
<Text color="red">Hello world</Text>
```
This would display as red, but if I don't have a red colour defined in my theme it is doing it based on the HTML colour, which undermines using the theme for consistency.

### Proposed Solution or API

A new flag for Chakra CLI, maybe `--strict-theme-tokens`.

When provided, the unions in the output wouldn't include `| (string & {})`.

At runtime, custom strings would still be allowed, but the typechecker would enforce correctness.

**Comment by Crizzooo** [comment]
We are using a custom chakra theme to support all front end repos at my company, and we'd like even stricter typing. 

:+1: to this idea. 

When overriding `fontFamily` we would like the `string | {}` removed so that the only valid values are the ones in `ThemeTypings['fonts']`.

Just confirming that our desired effect would align with this proposal so I don't create a duplicate ask, right @lexanth ?

**Comment by chrisregner** [comment]
I would add `textStyles` as well. Actually, it's not obvious why we wouldn't want to make all the theme properties strictly type.

**Comment by lexanth** [comment]
> Just confirming that our desired effect would align with this proposal so I don't create a duplicate ask, right @lexanth ?

> I would add textStyles as well. Actually, it's not obvious why we wouldn't want to make all the theme properties strictly typed.

Yes, I think the general request here is an about "strict" typing mode for everything.

I think the argument for non-strict is to match the runtime behaviour, which allows complete flexibility.

**Comment by lexanth** [comment]
An additional change that would be required is this type:
```
export type Union<T> = T | (string & {})

export type Token<
  CSSType,
  ThemeKey = unknown,
> = ThemeKey extends keyof ThemeTypings
  ? ResponsiveValue<Union<CSSType | ThemeTypings[ThemeKey]>>
  : ResponsiveValue<CSSType>
```

If we have theme typings for this key, we allow any of the theme keys, or any of the underlying CSS values, or any string.

I don't entirely understand why we'd want to be allowing any string in the theme key case, but not in the non-theme-key case. The non-theme-key case was changed in the PR which added type generation - https://github.com/chakra-ui/chakra-ui/pull/3092. Given that the theme types currently include `| (string & {})` anyway, it seems redundant, because `ThemeTypings[ThemeKey]` would already allow any string.

@TimKolberger does that sound right? Would a strict generation mode PR be acceptable?

**Comment by TimKolberger** [comment]
> Would a strict generation mode PR be acceptable?

Yes, I think this would be a great feature for codebases which rely on a strict design system. Love it 💖

The default typings should allow the user to define every valid css type and theme type - like today.
Only when the user generated a strict theme typing with the cli or with TS module augmentation it should allow only those theme token typings.

> I don't entirely understand why we'd want to be allowing any string in the theme key case, but not in the non-theme-key case. 

I am not sure either if this is a type duplication - I'd love to see a draft PR and then we iterate on it.

I think we should give this idea a spin 🚀

**Comment by lexanth** [comment]
Not stale - attached PR is approved

**Comment by lexanth** [comment]
@TimKolberger Anything I can do to get the attached PR moving forward?
